sigmastate/src/main/scala/sigmastate/serialization/ContractTemplateSerializer.scala

```diff
@@ -0,0 +1,144 @@
+  def serializeContractTemplate(contractTemplate: ContractTemplate): Array[Byte] = {
+    val res = contractTemplate.expressionTree match {
+      case Left(UnparsedContractTemplate(bytes, _)) => bytes.array
+      case _ =>
+        val w = SigmaSerializer.startWriter()
+        serializeHeaders(contractTemplate, w)
+        val expressionBytes = serializeWithoutSize(contractTemplate)
+        w.putUInt(expressionBytes.length)
+        w.putBytes(expressionBytes)
+        w.toBytes
+    }
+    res
+  }
+
+  /** Default deserialization of ContractTemplate (should be inverse to `serializeContractTemplate`).
+    * Doesn't apply any transformations to the values. */
+  def deserializeContractTemplate(bytes: Array[Byte]): ContractTemplate = {
+    val r = SigmaSerializer.startReader(bytes)
+    deserializeContractTemplate(r)
+  }
+
+  private[sigmastate] def deserializeContractTemplate(r: SigmaByteReader): ContractTemplate = {
+    val startPos = r.position
+    val treeVersion = r.getOption(r.getUByte().toByte)
+
+    val nameLength = r.getUInt().toInt
+    val name = new String(r.getBytes(nameLength), StandardCharsets.UTF_8)
+
+    val descriptionLength = r.getUInt().toInt
+    val description = new String(r.getBytes(descriptionLength), StandardCharsets.UTF_8)
+
+    val nConstants = r.getUInt().toInt
+    val constTypes: IndexedSeq[SType] = {
+      if (nConstants > 0) {
+        // HOTSPOT:: allocate new array only if it is not empty
+        val res = safeNewArray[SType](nConstants)
+        cfor(0)(_ < nConstants, _ + 1) { i =>
+          res(i) = TypeSerializer.deserialize(r)
+        }
+        res
+      } else {
+        SType.EmptySeq
+      }
+    }
+    val constValues: Option[IndexedSeq[Option[SType#WrappedType]]] = r.getOption((() => {
+      if (nConstants > 0) {
+        // HOTSPOT:: allocate new array only if it is not empty
+        val res = safeNewArray[Option[SType#WrappedType]](nConstants)
+        cfor(0)(_ < nConstants, _ + 1) { i =>
+          res(i) = r.getOption((() => DataSerializer.deserialize(constTypes(i), r))())
+        }
+        res
+      } else {
+        Array.empty[Option[SType#WrappedType]]
+      }
+    })())
+
+    val nParameters = r.getUInt().toInt
+    val parameters: IndexedSeq[Parameter] = {
+      if (nParameters > 0) {
+        val res = safeNewArray[Parameter](nParameters)
+        cfor(0)(_ < nParameters, _ + 1) { i =>
+          res(i) = ParameterSerializer.deserialize(r)
+        }
+        res
+      } else {
+        Parameter.EmptySeq
+      }
+    }
+
+    // TODO: Here we need to create the list of constants (IndexedSeq[Constant[T]]) and populate it in the constant
+    // store as r.constantStore so that the expressionTree which contains ConstantPlaceholder's can be deserialized.
+    // Unfortunately, the value part of the constant is not always known here and can be None here.
+//    constTypes.indices.map(i => DeserializationSigmaBuilder.mkConstant(null, constTypes(i)))
```

Comment:
@aslesarenko Need your input here. 

In order to deserialize the `expressionTree` of the ContractTemplate, we need to have all the constants defined in the `constantStore`. This is because the serialized `expressionTree` can have `ConstantPlaceholder` in the tree. The deserializer for the placeholder looks up the constant from the constantStore during the deserialization process - [code link](https://github.com/ScorexFoundation/sigmastate-interpreter/blob/dab64a3c8b4568f3e84adbb48842047c9c0f1f7f/sigmastate/src/main/scala/sigmastate/serialization/ConstantPlaceholderSerializer.scala#L21)

For some of the parameters, the value in `constValues` will be `None`. Hence, it is not possible to populate constants in the constantStore since we don't know the value of them.

Basically, we need to be able to do something like `DeserializationSigmaBuilder.mkConstant(null, constTypes(i))` i.e. set the value as `null` but I don't think that's possible.

Reply:
Done, added a `defaultOf` method.